### PR TITLE
Improved cache docs

### PR DESCRIPTION
**Pull request by manueldeprada**
Improved documentation for cache. Draft to be improved with the per-layer cache refactor.

EDIT: submitting this for review since the per-layer refactor is a thing on its own. Happy to merge this small improvement.

### CODE REVIEW

File: docs/source/en/cache_explanation.md [hunk 1]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
 
-When you use Transformers' [`Cache`] class, the self-attention module performs several critical steps to integrate past and present information.
+Thus, the basic interface of a KV cache takes a key and value tensor for the current token and returns the updated `K` and `V` tensors.
+
+```py
+new_K, new_V = cache.update(k_t, v_t, layer_idx)
+attn_output = attn_layer_idx_fn(q_t, new_K, new_V)
+```
+
+This is managed internally by the model's `forward` method. When you use Transformers' [`Cache`] class, the self-attention module performs several critical steps to integrate past and present information.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:76]:
```suggestion
When you use Transformers' [`Cache`] class, the self-attention module performs several critical steps to integrate past and present information.
```

File: docs/source/en/cache_explanation.md [hunk 2]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
 
 2. When the `forward` method is called iteratively, it's crucial that the attention mask shape matches the combined length of the past and current kv pairs. The attention mask should have the shape `(batch_size, past_kv_length + new_tokens_length)`. This is typically handled internally in [`~GenerationMixin.generate`], but if you want to implement your own generation loop with [`Cache`], keep this in mind! The attention mask should hold the past and current token values.
 
 3. It is also important to be aware of the `cache_position`. This is important if you want to reuse a prefilled [`Cache`] with the `forward` method because you have to pass a valid `cache_position` value. This indicates the input positions in a sequence. `cache_position` is unaffected by padding, and it always adds one more position for each token. For example, if a kv cache contains 10 tokens - regardless of pad tokens - the cache position for the next token should be `torch.tensor([10])`.
 
+## Data Storage Implementation
+
+The actual storage of key-value pairs varies between cache implementations. Let's look at the most basic implementation, `DynamicCache`:
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:86]:
```suggestion
## Cache storage implementation

The actual storage of key-value pairs varies between cache implementations. As an example, consider the [`DynamicCache`].
```

File: docs/source/en/cache_explanation.md [hunk 3]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+
+### DynamicCache Implementation
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:88]:
```suggestion
```

File: docs/source/en/cache_explanation.md [hunk 4]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+
+In `DynamicCache`, the key-value pairs are stored as two lists of tensors:
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:90]:
```suggestion
In [`DynamicCache`], the key-value pairs are stored as two lists of tensors. Each tensor in the lists have the shape `[batch_size, num_heads, seq_len, head_dim]`.
```

File: docs/source/en/cache_explanation.md [hunk 5]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+- `key_cache`: A list of tensors, one for each layer
+- `value_cache`: A list of tensors, one for each layer
+
+Each tensor in these lists has the shape `[batch_size, num_heads, seq_len, head_dim]`. When new tokens are processed:
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:94]:
```suggestion
When new tokens are processed:
```

File: docs/source/en/cache_explanation.md [hunk 6]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+
+1. For each layer, the new key and value states are concatenated with the existing cache:
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:96]:
```suggestion
1. For each layer, the new key and value states are concatenated with the existing cache.
```

File: docs/source/en/cache_explanation.md [hunk 7]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+```py
+self.key_cache[layer_idx] = torch.cat([self.key_cache[layer_idx], key_states], dim=-2)
+self.value_cache[layer_idx] = torch.cat([self.value_cache[layer_idx], value_states], dim=-2)
+```
+
+2. The cache grows dynamically as more tokens are processed, with the sequence length dimension (`seq_len`) increasing with each new token.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:102]:
```suggestion
2. The cache grows dynamically as more tokens are processed. The sequence length dimension (`seq_len`) increases with each new token.
```

File: docs/source/en/cache_explanation.md [hunk 8]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
+
+3. The cache maintains a count of seen tokens through `self._seen_tokens`, which is updated when the first layer processes a new token.
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:104]:
```suggestion
3. The cache maintains a count of seen tokens through `self._seen_tokens`. This is updated when the first layer processes a new token.
```

File: docs/source/en/cache_explanation.md [hunk 9]
```diff
@@ -72,10 +136,14 @@ for _ in range(max_new_tokens):
 print(tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0])
 "[INST] Hello, what's your name. [/INST]  Hello! My name is LLaMA,"
 ```
-
 ## Legacy cache format
 
-Before the [`Cache`] class, the cache used to be stored as a tuple of tuples of tensors. This format has is dynamic because it grows as text is generated, similar to [`DynamicCache`].
+Before the [`Cache`] class, the cache used to be stored as a tuple of tuples of tensors. This format is dynamic because it grows as text is generated, similar to [`DynamicCache`].
+
+The legacy format is essentially the same data structure but organized differently:
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:143]:
```suggestion
The legacy format is essentially the same data structure but organized differently.
```

File: docs/source/en/cache_explanation.md [hunk 10]
```diff
@@ -29,16 +29,80 @@ A key-value (KV) cache eliminates this inefficiency by storing kv pairs derived
...
 
+## Data Storage Implementation
+
+The actual storage of key-value pairs varies between cache implementations. Let's look at the most basic implementation, `DynamicCache`:
+
+### DynamicCache Implementation
+
+In `DynamicCache`, the key-value pairs are stored as two lists of tensors:
+- `key_cache`: A list of tensors, one for each layer
+- `value_cache`: A list of tensors, one for each layer
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:92]:
```suggestion
- `key_cache`: A list of tensors, one for each layer.
- `value_cache`: A list of tensors, one for each layer.
```

File: docs/source/en/cache_explanation.md [hunk 11]
```diff
@@ -72,10 +136,14 @@ for _ in range(max_new_tokens):
...
-
 ## Legacy cache format
 
-Before the [`Cache`] class, the cache used to be stored as a tuple of tuples of tensors. This format has is dynamic because it grows as text is generated, similar to [`DynamicCache`].
+Before the [`Cache`] class, the cache used to be stored as a tuple of tuples of tensors. This format is dynamic because it grows as text is generated, similar to [`DynamicCache`].
+
+The legacy format is essentially the same data structure but organized differently:
+- It's a tuple of tuples, where each inner tuple contains the key and value tensors for a layer
+- The tensors have the same shape `[batch_size, num_heads, seq_len, head_dim]`
+- The format is less flexible and doesn't support advanced features like quantization or offloading
```
Comment (stevhliu) [inline docs/source/en/cache_explanation.md:146]:
```suggestion
- It's a tuple of tuples, where each inner tuple contains the key and value tensors for a layer.
- The tensors have the same shape `[batch_size, num_heads, seq_len, head_dim]`.
- The format is less flexible and doesn't support features like quantization or offloading.
```

**Review by stevhliu — APPROVED** [review]
Super nice, thanks for your improvements! 🎉
